odl/contrib/pytorch/layer.py

```diff
@@ -0,0 +1,201 @@
+            Variable holding the result of applying the Jacobian to
+            ``grad_output``. See Notes for details.
+
+        Examples
+        --------
+        Compute the Jacobian adjoint of the matrix operator, which is the
+        operator of the transposed matrix. We compose with the ``sum``
+        functional to be able to evaluate ``grad``:
+
+        >>> space = odl.rn(3)
+        >>> matrix = np.array([[1, 0, 1],
+        ...                    [0, 1, 1]], dtype=float)
+        >>> op = odl.MatrixOperator(matrix, domain=space)
+        >>> matrix_op = TorchOperator(op)
+        >>> x_elem = space.element([1, 2, 3])
+        >>> x = torch.from_numpy(x_elem.asarray())
+        >>> x_var = torch.autograd.Variable(x, requires_grad=True)
+        >>> op_x_var = matrix_op(x_var)
+        >>> cost = op_x_var.sum()
+        >>> cost.backward()
+        >>> x_var.grad
+
+
+        Compute the gradient of a custom functional:
+
+        >>> space = odl.rn(3)
+        >>> functional = odl.solvers.L2NormSquared(space)
+        >>> func_op = TorchOperator(functional)
+        >>> x_elem = space.element([1, 2, 3])
+        >>> x = torch.from_numpy(x_elem.asarray())
+        >>> x_var = torch.autograd.Variable(x, requires_grad=True)
+        >>> func_x_var = func_op(x_var)
+        >>> func_x_var
+        Variable containing:
+         14
+        [torch.DoubleTensor of size 1]
+
+        >>> func_x_var.backward()
+        >>> x_var.grad  # Should be 2 * x
+        Variable containing:
+         2
+         4
+         6
+        [torch.DoubleTensor of size 3]
+
+        Notes
+        -----
+        This method applies the contribution of this node, i.e., the
+        transpose of the Jacobian of its outputs with respect to its inputs,
+        to the gradients of some cost function with respect to the outputs
+        of this node.
+
+        Example: Assume that this node computes :math:`x \mapsto f(x)`, and
+        its output is connected to a cost function that computes
+        :math:`y --> C(y)`. Here, :math:`x`, :math:`y` and :math:`f(x)` are
+        tensor variables and :math:`C(y)` is a scalar variable.
+        In ODL language, what ``backward`` should compute is
+
+            .. math::
+                \\nabla(C \circ f)(x) = f'(x)^*\\big(\\nabla C (f(x))\\big)
+
+        according to the chain rule. In ODL code, this corresponds to ::
+
+            f.derivative(x).adjoint(C.gradient(f(x))).
+
+        Then, the parameter ``grad_output`` contains a single tensor
+        variable ``y`` that stands for :math:`\\nabla C(f(x))`. Thus,
+        ``backward`` boils down to computing ``[f'(x)^*(y)]``, where
+        ``grad_output == [y]`` and ``inputs == [x]``.
+        """
+        # TODO: scaling
+        input, = self.saved_variables
+        grad = None
+
+        if self.needs_input_grad[0]:
+            input_arr = input.data.numpy()
+            # TODO: something wrong here, getting strides=(0,) sometimes
```

Comment:
In what cases are you getting strides=(0,)?

Reply:
When using the torch builtin `sum()` in particular, I guess. It may be a very memory-friendly way of storing an array of all same values, but it's not compatible with `matrix.dot()` using BLAS. Perhaps `MatrixOperator` needs a check for that.

Reply:
In particular, it may be a feature, not a bug, and we have to deal with it.

Reply:
It's a Numpy issue, and the fix is [here](https://github.com/numpy/numpy/pull/9177). In the meanwhile I guess we have to copy in that case. Or maybe always, since in-place mutation (like `arr *= 2`) yields wrong results.

Reply:
For the records, the `broadcast_to` function also creates a zero-stride array:
```python
>>> np.broadcast_to(np.ones((3, 1)), (3, 4)).strides
(8, 0)
```

Reply:
Overall 0 stride arrays are not really valid datacontainers for ODL since they don't work with in place operations. Perhaps we should disallow them.

Reply:
True. In the cases when Numpy produces them (like `broadcast_to`), the arrays are usually read-only, for a good reason. So we should stay away from them.